### ISSUE TITLE
Unable to extract an accuracy/confidence metric from the OCR

### ISSUE THREAD

**Issue by PantelisDeveloping** [Ucas-HaoranWei/GOT-OCR2.0#148]
Dear Authors,

Thank you for this lovely project. Please i try to implement your OCR library but i am unable to implement a confidence metric of the detected OCR accuracy. If possible kindly suggest an approach or implement it. In case it has been please let me know thank you so much.

The following is the code i am working on.

`from transformers import AutoModel, AutoTokenizer
import torch
from PIL import Image
import os
import numpy as np

def perform_ocr_with_confidence(image_path, model_name='ucaslcl/GOT-OCR2_0'):
    """
    Perform OCR on an image and calculate confidence scores
    
    Args:
        image_path (str): Path to the image file
        model_name (str): Name of the model to use
    
    Returns:
        dict: OCR results and confidence metrics
    """
    if not os.path.exists(image_path):
        raise FileNotFoundError(f"Image file not found: {image_path}")

    try:
        # Initialize tokenizer and model
        tokenizer = AutoTokenizer.from_pretrained(
            model_name, 
            trust_remote_code=True
        )
        
        model = AutoModel.from_pretrained(
            model_name,
            trust_remote_code=True,
            low_cpu_mem_usage=True,
            device_map='cuda',
            use_safetensors=True,
            pad_token_id=tokenizer.eos_token_id
        )
        
        # Set model to evaluation mode
        model = model.eval().cuda()
        
        # Process image and get output with logits
        with torch.no_grad():
            # Get both text output and logits
            outputs = model.chat(
                tokenizer,
                image_path,
                ocr_type='ocr',
                output_attentions=True,  # Get attention weights
                output_hidden_states=True  # Get hidden states
            )
            
            # Access model's hidden states for confidence calculation
            if hasattr(model, 'transformer'):
                hidden_states = model.transformer.last_hidden_state
                
                # Calculate confidence based on hidden state magnitudes
                # Higher magnitudes generally indicate higher confidence
                confidence_scores = torch.norm(hidden_states, dim=-1).mean().item()
                
                # Normalize to 0-1 range
                confidence_normalized = 1 / (1 + np.exp(-confidence_scores))
                
                # Calculate character-level confidences
                char_confidences = torch.norm(hidden_states, dim=-1).tolist()
                
                return {
                    'text': outputs,
                    'overall_confidence': float(confidence_normalized),
                    'char_confidences': char_confidences,
                    'metadata': {
                        'model': model_name,
                        'image': image_path
                    }
                }
            else:
                # Fallback if we can't access hidden states
                return {
                    'text': outputs,
                    'overall_confidence': None,
                    'warning': 'Could not calculate confidence scores',
                    'metadata': {
                        'model': model_name,
                        'image': image_path
                    }
                }
            
    except Exception as e:
        raise Exception(f"Error performing OCR: {str(e)}")

def print_results_with_confidence(results):
    """
    Print OCR results with confidence scores in a readable format
    """
    print("\nOCR Results:")
    print("="* 50)
    print(f"Detected Text: {results['text']}")
    print("-" * 50)
    
    if results['overall_confidence'] is not None:
        confidence_percentage = results['overall_confidence'] * 100
        print(f"Overall Confidence: {confidence_percentage:.2f}%")
        
        # Print confidence indicator bar
        bar_length = 40
        filled_length = int(bar_length * results['overall_confidence'])
        bar = '█' * filled_length + '░' * (bar_length - filled_length)
        print(f"Confidence Level: |{bar}| {confidence_percentage:.1f}%")
        
        if confidence_percentage < 50:
            print("Warning: Low confidence detection - results may need verification")
    else:
        print("Confidence scores not available")
    
    print("-" * 50)
    print(f"Model Used: {results['metadata']['model']}")
    print(f"Image Processed: {results['metadata']['image']}")

# Usage example
if __name__ == "__main__":
    image_file = 'my/image/path.png'
    
    try:
        results = perform_ocr_with_confidence(image_file)
        print_results_with_confidence(results)
            
    except Exception as e:
        print(f"Error: {str(e)}")`

**Comment by Ucas-HaoranWei** [comment]
Do you mean the benchmark？ https://github.com/Ucas-HaoranWei/GOT-OCR2.0/tree/main/GOT-OCR-2.0-master/GOT/eval
and you can download the benchmark in https://drive.google.com/drive/folders/1OdDtsJ8bFJYlNUzCQG4hRkUL6V-qBQaN

**Comment by ayumiymk** [comment]
I also have the same question: how can we obtain the probability of the OCR results?

**Comment by PantelisDeveloping** [comment]
Hello, thank you for your prompt response,

I don't mean the benchmarks themselves. Sometimes when I perform OCR on various documents in some cases I fail to obtain the result accurately. I would like to implement this: "For a given image we performed OCR with an accuracy of ***%". How can we measure the accuracy of the OCR ourselves for a given document image? Thank you for your understanding.

**Comment by Ucas-HaoranWei** [comment]
> 希望得到类似PaddleOCR的置信度

那是检测的置信度吧，不是ocr（识别）的置信度吧？

**Comment by Ucas-HaoranWei** [comment]
> Hello, thank you for your prompt response,
> 
> I don't mean the benchmarks themselves. Sometimes when I perform OCR on various documents in some cases I fail to obtain the result accurately. I would like to implement this: "For a given image we performed OCR with an accuracy of ***%". How can we measure the accuracy of the OCR ourselves for a given document image? Thank you for your understanding.

It is actually the detection accuracy？

**Comment by futureisatyourhand** [comment]
对于类似的问题，我有个疑问，通过构建合适的prompt，能否基于GOT-OCR 2.0微调实施文本定位功能？作者有相应的尝试嘛
